src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs

```diff
@@ -536,39 +550,47 @@ private void RaiseFunctionMetricEvent(FunctionStartedEvent runningFunctionInfo,
 
             private void RaiseMetricsPerFunctionEvent()
             {
-                List<FunctionMetrics> metricsEventsList = GetMetricsQueueSnapshot();
-
-                var aggregatedEventsPerFunction = from item in metricsEventsList
-                                                  group item by item.FunctionName into functionGroups
-                                                  select new
-                                                  {
-                                                      FunctionName = functionGroups.Key,
-                                                      StartedCount = Convert.ToUInt64(functionGroups.Count(x => x.ExecutionStage == ExecutionStage.Started)),
-                                                      FailedCount = Convert.ToUInt64(functionGroups.Count(x => x.ExecutionStage == ExecutionStage.Failed)),
-                                                      SucceededCount = Convert.ToUInt64(functionGroups.Count(x => x.ExecutionStage == ExecutionStage.Succeeded)),
-                                                      TotalExectionTimeInMs = Convert.ToUInt64(functionGroups.Sum(x => Convert.ToDecimal(x.ExecutionTimeInMS)))
-                                                  };
-
-                foreach (var functionEvent in aggregatedEventsPerFunction)
+                // Snapshot existing metrics by swapping out the collection we're building.
+                // New metrics rolling in will go to the new collection while we aggregate the snapshot here.
+                var summarySnapshot = Interlocked.Exchange(ref _functionMetricsSummary, new ConcurrentDictionary<string, Lazy<FunctionMetricSummary>>());
+
+                foreach (var function in summarySnapshot)
                 {
-                    MetricsEventGenerator.LogFunctionExecutionAggregateEvent(_appServiceOptions.AppName, functionEvent.FunctionName, (long)functionEvent.TotalExectionTimeInMs, (long)functionEvent.StartedCount, (long)functionEvent.SucceededCount, (long)functionEvent.FailedCount);
+                    var summary = function.Value.Value;
+                    MetricsEventGenerator.LogFunctionExecutionAggregateEvent(_appServiceOptions.AppName, function.Key, summary.TotalExectionTimeInMs, summary.StartedCount, summary.SucceededCount, summary.FailedCount);
                 }
             }
 
-            private List<FunctionMetrics> GetMetricsQueueSnapshot()
+            /// <summary>
+            /// This evil mutable struct is for keeping metrics *as we go*, rather than collecting events and summarizing on the timer interval.
+            /// In the aggregate: it's a much faster and lower allocation way of tracking the same numbers.
+            /// </summary>
+            private class FunctionMetricSummary
```

Comment:
Comment says this is a struct but it's a class?

Reply:
Ah yes this evolved due to the StyleCop issue so we wouldn't delay - will update comment!